lib/dataClasses/seed_response.dart

```diff
@@ -0,0 +1,43 @@
+import 'dart:convert';
+
+/// The SeedResponse class represents the data from a response from the seed
+/// server. It accepts input for deserialization. The factory constructors will
+/// not pass empty or default versions of the class. Rather, they throw
+/// expcetions if the data passed in are invalid.
+class SeedResponse {
+  final String seed;
+  final DateTime expiration;
+
+  SeedResponse({
+    required this.seed,
+    required this.expiration,
+  });
+
+  /// fromJsonString parses all values from the JSON data object passed in and
+  /// attempts to create a SeedResponse object. A couplt things to note:
+  /// This constructor throws exceptions if the data is invalid.
+  /// I didn't use a generator because the Flutter page itself states:
+  /// 'Use manual serialization for smaller projects'
+  /// https://flutter.dev/docs/development/data-and-backend/json
+  factory SeedResponse.fromJsonString(String jsonStr) {
+    var rawJson = jsonDecode(jsonStr);
+
+    const errMsg = 'Invalid JSON received from server';
+    if (rawJson is! Map) {
+      throw Exception(errMsg);
+    }
+
+    if (rawJson['seed'] is! String || rawJson['expires_at'] is! String) {
+      throw Exception(errMsg);
+    }
+
+    DateTime exp;
+    try {
+      exp = DateTime.parse(rawJson['expires_at']);
+    } catch (e) {
+      throw Exception(errMsg);
```

Comment:
**question:** is the error message here still accurate to the error we're checking for or throwing for?

Reply:
Same response from above:

Yes. I wrote the fromJsonString value to only accept the values from an API if:

- the JSON string represents a Map
- the seed key represents a string
- the expires_at key represents a string
- the expires_at key is parseable by DateTime.parse

If it doesn't satisfy all of the above, I consider the JSON string invalid.

In many situations, I usually handle invalid data by using default values, but I read the constraints of the program to required the above.